src/shared/utilities/vsCodeUtils.ts

```diff
@@ -0,0 +1,107 @@
+/*!
+ * Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+'use strict'
+import * as vscode from 'vscode'
+import * as nls from 'vscode-nls'
+export const localize = nls.loadMessageBundle()
+
+import { ErrorOrString, getLogger, Logger, LogLevel } from '../logger'
+
+// ------- Experimental combined output channel & traditional logger  --------------
+
+interface LogParams {
+    args: ErrorOrString[],
+    channel: vscode.OutputChannel,
+    level: LogLevel,
+    logger: Logger,
+    msg: string,
+    nlsKey: string,
+}
+
+// TODO: Leverage 3rd party code: https://raw.githubusercontent.com/Microsoft/vscode-nls/master/src/main.ts
+// function format(template: string, ...templateParams: any[]): string {
+// .......
+// }
+
+function log({ args = [], channel, level, msg, nlsKey, logger }: LogParams): void {
+    if (level === 'error') {
+        channel.show(true)
+    }
+    const [arg0, ...restArgs] = args
+
+    if (arg0 && arg0 instanceof Error) {
+        // TODO: Define standard/strategy. Sometimes err.message is part of msg. Probably shouldn't be.
+        channel.appendLine(
+            localize(nlsKey, msg, arg0.message, ...restArgs)
+        )
+        logger[level](
+            localize(nlsKey, msg, arg0.message, ...restArgs)
+            // TODO: Use english if/when we get multi lang support
+            // format(msg, arg0.message, ...restArgs)
+        )
+        if (arg0.stack) {
+            logger[level](arg0.stack)
+        }
+
+    } else {
+        channel.appendLine(
+            localize(nlsKey, msg, ...args)
+        )
+        logger[level](
+            // TODO: Use english if/when we get multi lang support
+            localize(nlsKey, msg, ...args)
+            // format(msg, ...args)
+        )
+    }
+}
+
+/**
+ * Wrapper around normal logger that writes to output channel and normal logs.
+ * Avoids making two log statements when writing to output channel and improves consistency
+ * @param channelName: Name of the output channel to write to
+ */
+export function getChannelLogger(channel: vscode.OutputChannel, logger: Logger = getLogger()) {
+    return {
```

Comment:
Nit: What do you think of:

```typescript
const sharedArgs = {
    args,
    channel,
    logger,
    msg,
    nlsKey
}

return {
    debug: (nlsKey: string, msg: string, ...args: ErrorOrString[]) => log({
        level: 'debug',
        ...sharedArgs
    }),
    // etc.
}
```

Reply:
It sure looks good. I wish it worked.

Reply:
Can you elaborate on why it doesn't work?

Reply:
`nlsKey`, `msg` and `args` are passed to the `debug` function call and are not available above where your `sharedArgs` is declared.